clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2682,19 +2696,46 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   }
   O << "};\n\n";
 
+  O << "// array representing kernel parameters used in all kernels defined in "
+       "the\n";
+  O << "// corresponding source\n";
+  O << "static constexpr\n";
+  O << "const bool param_omit_table[] = {\n";
+  O << "  // OMIT_TABLE_BEGIN\n";
+  for (const KernelDesc &K : KernelDescs) {
+    O << "  //--- " << K.Name << "\n";
+    O << "  ";
+    for (const KernelParamDesc &P : K.Params) 
+      for (unsigned J = 0; J < P.NumOpenCLParams; J++)
+        O << "false, ";
+    O << "\n";
+  }
+  O << "  // OMIT_TABLE_END\n";
+  O << "  }; \n\n";
+
   O << "// array representing signatures of all kernels defined in the\n";
   O << "// corresponding source\n";
   O << "static constexpr\n";
   O << "const kernel_param_desc_t kernel_signatures[] = {\n";
 
-  for (unsigned I = 0; I < KernelDescs.size(); I++) {
-    auto &K = KernelDescs[I];
+  unsigned CurIndex = 0;
+  for (const KernelDesc &K : KernelDescs) {
     O << "  //--- " << K.Name << "\n";
-
-    for (const auto &P : K.Params) {
+    for (const KernelParamDesc &P : K.Params) {
       std::string TyStr = paramKind2Str(P.Kind);
       O << "  { kernel_param_kind_t::" << TyStr << ", ";
-      O << P.Info << ", " << P.Offset << " },\n";
+      O << P.Info << ", " << P.Offset << ", ";
+      for (unsigned X = 0; X < P.NumOpenCLParams; X++) {
```

Comment:
Sorry to change my mind here...

Pull the X== 0 case out of the for-loop (since we know it is printed always, and it would otherwise be an error to have it not print).  Then, the value in the loop (from unsigned X = 1; X < P.NumOpenCLParmas) can always prepend the bar.

ALSO, the parens around the single case aren't necessary, so it doesn't need that.